File: stock_search/data_sources/stockanalysis/exa_fallback.py
"""Generic Exa-backed fallback helpers for StockAnalysis pages."""

from __future__ import annotations

from collections.abc import Callable
import logging

from llm_harness.clients import ExaAgent

logger = logging.getLogger(__name__)


def invoke_stockanalysis_search[MODEL_TYPE](
    *,
    output_schema: type[MODEL_TYPE],
    system_prompt_template: str,
    query: str,
    prompt_values: dict[str, str],
) -> MODEL_TYPE:
    """Run a structured Exa search using a StockAnalysis-oriented prompt."""
    agent = ExaAgent(
        system_prompt=system_prompt_template.format(**prompt_values),
        output_schema=output_schema,
    )
    return agent.invoke(query)


def invoke_stockanalysis_search_or_default[MODEL_TYPE](
    *,
    output_schema: type[MODEL_TYPE],
    system_prompt_template: str,
    query: str,
    prompt_values: dict[str, str],
    default_factory: Callable[[], MODEL_TYPE],
    error_message: str,
) -> MODEL_TYPE:
    """Run a structured Exa search and return an empty model on failure."""
    try:
        result = invoke_stockanalysis_search(
            output_schema=output_schema,
            system_prompt_template=system_prompt_template,
            query=query,
            prompt_values=prompt_values,
        )
        if isinstance(result, output_schema):
            return result
    except Exception:
        logger.exception(error_message, prompt_values.get("ticker", "unknown"))
    return default_factory()

File: src/stock-search/data-sources/stockanalysis/exa-fallback.ts
/** Generic Exa-backed fallback helpers for StockAnalysis pages. */

import type { z, ZodType } from "zod";

type ExaAgentCtor = new <T extends ZodType>(
	systemPrompt: string,
	outputSchema: T,
) => {
	invoke(query: string): Promise<z.output<T>>;
};

async function loadExaAgent(): Promise<ExaAgentCtor> {
	const importer = new Function(
		"return import('../../../../../llm-harness-js/src/agents/index.js')",
	) as () => Promise<{ ExaAgent: ExaAgentCtor }>;
	const module = await importer();
	return module.ExaAgent;
}

/** Run a structured Exa search using a StockAnalysis-oriented prompt. */
export async function invokeStockanalysisSearch<T extends ZodType>({
	outputSchema,
	systemPromptTemplate,
	query,
	promptValues,
}: {
	outputSchema: T;
	systemPromptTemplate: string;
	query: string;
	promptValues: Record<string, string>;
}): Promise<z.output<T>> {
	const ExaAgent = await loadExaAgent();
	const agent = new ExaAgent(
		systemPromptTemplate.replace(
			/\{([^}]+)\}/g,
			(_match, key: string) => promptValues[key] ?? "",
		),
		outputSchema,
	);
	return agent.invoke(query);
}

/** Run a structured Exa search and return a default value on failure. */
export async function invokeStockanalysisSearchOrDefault<T extends ZodType>({
	outputSchema,
	systemPromptTemplate,
	query,
	promptValues,
	defaultFactory,
}: {
	outputSchema: T;
	systemPromptTemplate: string;
	query: string;
	promptValues: Record<string, string>;
	defaultFactory: () => z.output<T>;
}): Promise<z.output<T>> {
	try {
		return await invokeStockanalysisSearch({
			outputSchema,
			systemPromptTemplate,
			query,
			promptValues,
		});
	} catch {
		return defaultFactory();
	}
}
